### ISSUE TITLE
Is the count correct?

### ISSUE THREAD

**Issue by ioquatix** [mperham/connection_pool#167]
https://github.com/mperham/connection_pool/blob/428c06f34209ee7a99f1ace5b96e567841c00d1c/lib/connection_pool.rb#L122

It seems to me, we increment the count, and then perform an operation to get a connection, which may time out.

It seems the invariant of `checkout` can be violated, i.e. it increments the count, even though it fails to check out a connection.

This is actually, somewhat suprisingly, handled correctly in `with` because the timeout interrupt is deferred until within the `begin` block which has `ensure checkout` which correctly decrements the count, even in the case of an exception. But during that time, the actual count seems incorrect.

It would appear this makes it tricky/unsafe to use `checkout` and `checkin` in general, since the count invariant can be violated.

**Comment by mperham** [comment]
Yeah, I can see some shady edge cases. I wonder if we can write some tests which surface the bad behavior.

**Comment by ioquatix** [comment]
I tried to write some but ran out of time.

I think anything that calls checkout with a timeout can fail, i.e. it's very hard to write safe code that uses `checkout` and `checkin` directly.
